src/models/room-state.ts

```diff
@@ -43,7 +43,16 @@ export interface IMarkerFoundOptions {
      */
     timelineWasEmpty?: boolean;
 }
-
+export interface ISetStateEventsOptions {
+    /** Whether the event is inserted at the start of the timeline
+     * or at the end. This is relevant for doing a sanity check:
+     * "Is the event id of the added event the same as the replaces_state id of the current event"
+     * We need to do this because sync sometimes feeds previous state events.
+     * If set to true the sanity check is inverted
+     * "Is the event id of the current event the same as the replaces_state id of the added event"
+     */
```

Comment:
I don't think this makes sense, as a RoomState is only ever in one direction. This sounds like if toStartOfTimeline=false then you should always drop the state event as they're never live. It should be a property on a RoomState if required

Reply:
That is a good observation. The state itself might be the right location to store if it's a start state or an end state.
It should never change and also not be relevant outside the state object it seems logical to belong to the state.

I have trouble understanding the exact definition of start state anyway however:

If we have a state that is only set at the beginning of the room, events from a time-line in a specific interval cannot be enough to compute the start state because the start state needs all state events BEFORE the start of the time-line but the time-line only includes events after the "start".

This is slightly off topic but i still wanted to being it up. Feel free to DM me about this or point me to where i can lookup how start state is defined.

Reply:
as I understand it, it is the state at the start of the timeline, and it walks backwards as you backpaginate and load older events which undo the more recent state events.

Reply:
That would also be how i understand start state. But this is not what we are doing.
We use the content to set the state keys. What we would need to do is set the state with the previous_content field.

| = start of timeline
|| = end of timeline
--- = time steps
XV> = state event changing the state (the content of the state event is the current state from that time on (X is an arbitrary label for the state event (f.ex. a member event) X is the type and V is the value

X0>-----|X1>----X2>----||
Here we have an example where the state is changed with the first event in the timeline. But the start imo is before that  event. So the state at the start is X0.

This becomes more clear when adding another type:
X0>--A0>--|A1>-A2>-X1>--||
At the start of the timeline we have X0 A0 but definitly not X1 which would be the first X type state event.

If we use the previous_content we have <XV as the previous content.

<X0>--<A0>--|<A1>-<A2>-<X1>--||

Then <A1 & <X1 is the same as X0> & A0> and the correct start state.

Reply:
I suggest raising this to the wider team as the code & intent predates me

Reply:
sure. Sorry for escalating this but I always found the startState to be arguably the correct name for that variable the way its implemented and this seemed to be relevant for this discussion. Somewhat relevant, you comment is just about where to store start/end.
What I do like about making it part of the setStateEvents property is that is makes it testable in a more isolated env.

Reply:
To get back to the topic what do you think about the more encapsulated testability with the current approach?